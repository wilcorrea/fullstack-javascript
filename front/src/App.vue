<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
