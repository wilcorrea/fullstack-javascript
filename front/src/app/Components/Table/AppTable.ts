import { CreateElement, VNode } from 'vue'
import AppTableBody from 'src/app/Components/Table/AppTableBodyMixin'
import AppTableHead from 'src/app/Components/Table/AppTableMixinHead'
import AppTableValue from 'src/app/Components/Table/AppTableMixinValue'
import AppTableStyle from 'src/app/Components/Table/AppTableMixinStyle'
import AppTableEvent from './AppTableEvent'
import { block } from './index'

/**
 */
export default {
  /**
   */
  name: 'AppTable',
  /**
   */
  mixins: [
    AppTableHead, AppTableBody, AppTableValue, AppTableStyle, AppTableEvent
  ],
  /**
   */
  props: {
    columns: {
      type: Array,
      required: true
    },
    records: {
      type: Array,
      default: () => []
    },
    loading: {
      type: Boolean,
      default: false
    },
    loader: {
      type: String,
      default: '<div class="ld ld-ring ld-vortex-in"></div>'
    },
    pressHold: {
      type: Boolean,
      default: true
    }
  },
  /**
   */
  methods: {
    /**
     * @param {CreateElement} h
     * @returns {VNode | undefined}
     */
    table (h: CreateElement): VNode | undefined {
      // @ts-ignore
      if (!this.columns) {
        return
      }
      const data = { class: block }
      const children = [
        this.tableContainer(h),
        this.tableShadow(h)
      ]

      return h('div', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableContainer (h: CreateElement): VNode {
      const root = `${block}__container`
      // @ts-ignore
      const loading = this.loading ? `${root}--loading` : ''
      const data = { class: [root, loading], attrs: { move: true } }
      // @ts-ignore
      const table = [this.thead(h, this.columns), this.tbody(h, this.columns, this.records)]
      const children = [
        h('table', table),
        this.tableLoading(h)
      ]
      return h('div', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableLoading (h: CreateElement): VNode {
      // @ts-ignore
      const visible = this.loading ? `${block}__loading--visible` : ''
      // @ts-ignore
      const html = `<div class="${block}__loading-container">${this.loader}</div>`
      const data = {
        class: [`${block}__loading`, visible],
        domProps: { innerHTML: html }
      }
      return h('div', data)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableShadow (h: CreateElement): VNode {
      // @ts-ignore
      const loading = this.loading ? `${block}__shadow--loading` : ''
      return h('div', { class: [`${block}__shadow`, loading] })
    }
  },
  /**
   * @param {CreateElement} h
   * @returns {VNode}
   */
  render (h: CreateElement): VNode {
    // @ts-ignore
    return this.table(h)
  }
}
