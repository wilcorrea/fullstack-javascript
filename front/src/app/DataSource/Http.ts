import { AxiosInstance, AxiosRequestConfig } from 'axios'

/**
 * @type {Http}
 */
export default class Http {
  /**
   * @type {string}
   */
  path = '/'

  /**
   * @type {{}}
   */
  static mocks = {}

  /**
   * @type {AxiosInstance}
   */
  private client: AxiosInstance

  /**
   * @type {Http}
   */
  private static single: Http

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static instance (client?: AxiosInstance) {
    if (!this.single) {
      // @ts-ignore
      this.single = new this(client)
    }
    return this.single
  }

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static build (client?: AxiosInstance) {
    // @ts-ignore
    return new this(client)
  }

  /**
   * @param {AxiosInstance} client
   */
  constructor (client: AxiosInstance) {
    this.client = client
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  post (url: string = '', data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.post(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  get (url: string = '', config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.get(url, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  head (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.head(url, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  put (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.put(url, data, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  patch (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.patch(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  delete (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.delete(url, config)
  }

  /**
   * @param {string} path
   * @param value
   */
  static mockRegister (path: string, value: any = null): void {
    // @ts-ignore
    Http.mocks[path] = value
  }

  /**
   * @param {string} path
   * @returns {any}
   */
  static mockRecover (path: string): any {
    // @ts-ignore
    return Http.mocks[path]
  }
}
