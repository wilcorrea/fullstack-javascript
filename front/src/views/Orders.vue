<template>
  <div class="app-layout__page Orders">
    <h1>{{ $lang('views.orders.title') }}</h1>
    <!-- the table is all in AppTable -->
    <!-- @see front/src/app/Components/Table/AppTable.ts -->
    <app-table v-bind="table" />
    <!-- pagination handle with offset and limit -->
    <!-- emits @paginationUpdateOffset when offset a new fetch is requested -->
    <!-- also emits @paginationToTop when the button is clicked -->
    <!-- @see front/src/app/Components/Pagination/AppPagination.ts -->
    <app-pagination
      v-bind="{ offset, limit }"
      @updateOffset="paginationUpdateOffset"
      @toTop="paginationToTop"
    />
  </div>
</template>

<script lang="js">
import OrderSchema from 'src/domains/Order/OrderSchema'
import OrderService from 'src/domains/Order/OrderService'
import { scrollDown, scrollUp } from 'src/app/Util/ui'

/**
 * @component {Orders}
 */
export default {
  /**
   */
  name: 'Orders',
  /**
   */
  data: () => ({
    // table states
    table: {
      columns: [],
      records: [],
      loading: true
    },
    // pagination controls
    offset: 0,
    limit: 5
  }),
  /**
   */
  methods: {
    /**
     * Method that use service to get the data
     * It starts the loading state
     */
    fetchRecords () {
      this.loadingShow()
      OrderService
        .instance()
        .search(this.offset, this.limit)
        .then(this.assignRecords)
        .catch(this.loadingHide)
    },
    /**
     * Assign records will update the table records
     * also will scroll down and finish loading state
     * @param {Object} result
     */
    assignRecords (result) {
      this.table.records.push(...result.orders)
      this.$nextTick(() => window.setTimeout(this.assignRecordsFinish, 200))
    },
    /**
     * Finally scroll down and hide the loading
     * It runs in a lazy way to don't overload the browser
     */
    assignRecordsFinish () {
      // don't scroll the first fetch
      if (this.offset !== 0) {
        scrollDown(this.$root.mainLayout)
      }
      this.loadingHide()
    },
    /**
     * Fired when the button is clicked will call a new fetch
     * @param {Object} options
     */
    paginationUpdateOffset (options) {
      const { offset } = options
      this.offset = offset
      this.fetchRecords()
    },
    /**
     * Scroll the page up
     */
    paginationToTop () {
      scrollUp(this.$root.mainLayout)
    },
    /**
     * Just show loading
     */
    loadingShow () {
      this.table.loading = true
    },
    /**
     * Just hide loading
     */
    loadingHide () {
      this.table.loading = false
    }
  },
  /**
   */
  created () {
    // hidrate the columns with schema definition
    this.table.columns = OrderSchema.instance().columns()
  },
  /**
   */
  mounted () {
    // perform the first fetch
    this.fetchRecords()
  }
}
</script>

<style lang="stylus">

</style>
