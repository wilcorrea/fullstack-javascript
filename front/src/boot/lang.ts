import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
