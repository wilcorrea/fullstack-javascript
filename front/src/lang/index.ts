import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
