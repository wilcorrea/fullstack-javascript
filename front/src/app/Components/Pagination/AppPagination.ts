import { CreateElement, VNode } from 'vue'
import $lang from 'src/lang'

/**
 * @type {string}
 */
const block = 'app-pagination'

/**
 * @component {AppPagination}
 */
export default {
  /**
   */
  name: 'AppPagination',
  /**
   */
  props: {
    offset: {
      type: Number,
      required: true
    },
    limit: {
      type: Number,
      required: true
    }
  },
  data: () => ({
    page: 0
  }),
  /**
   */
  methods: {
    /**
     * @returns {number}
     */
    nextPage (): number {
      // @ts-ignore
      return this.limit * ++this.page
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    renderButtonAppend (h: CreateElement): VNode {
      const data = {
        class: `${block}__append`,
        on: {
          click: () => {
            // @ts-ignore
            this.$emit('updateOffset', { offset: this.nextPage() })
          }
        }
      }
      const children = $lang('app.components.pagination.more')

      return h('button', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    renderButtonToTop (h: CreateElement): VNode {
      const data = {
        class: `${block}__to-top`,
        on: {
          click: () => {
            // @ts-ignore
            this.$emit('toTop')
          }
        }
      }
      const children = [h('app-icon', { props: { name: 'arrow_upward' } })]
      return h('button', data, children)
    }
  },
  /**
   * @param {CreateElement} h
   * @returns {VNode}
   */
  render (h: CreateElement): VNode {
    const data = { class: block }
    // @ts-ignore
    const children: Array<VNode> = [this.renderButtonAppend(h), this.renderButtonToTop(h)]
    return h('div', data, children)
  }
}
