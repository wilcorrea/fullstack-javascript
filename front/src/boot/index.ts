import lang from './lang'
import browse from './browse'
import util from './util'
import './polyfill'

/**
 * @type {{install: Function}}
 */
export default {
  install (Vue: any) {
    lang(Vue)
    browse(Vue)
    util(Vue)
  }
}
