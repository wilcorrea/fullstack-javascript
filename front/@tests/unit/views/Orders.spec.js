import { shallowMount } from '@vue/test-utils'
import Orders from 'src/views/Orders.vue'
import $lang from 'src/lang'

import AppTable from 'src/app/Components/Table/AppTable'
import AppPagination from 'src/app/Components/Pagination/AppPagination'
import Http from 'src/app/DataSource/Http'

describe('Orders', () => {
  const mock = { orders: [{ id: 1 }] }
  const offset = 0
  const limit = 5
  Http.mockRegister(`/orders/search/offset=${offset}/limit=${limit}`, mock)

  const wrapper = shallowMount(
    Orders,
    { components: { AppTable, AppPagination }, mocks: { $lang } }
  )

  /**
   * @test
   */
  it('check component markup', () => {
    expect(wrapper.html()).toContain(`<h1>${$lang('views.orders.title')}</h1>`)
  })

  /**
   * @test
   */
  it('method assignRecords', () => {
    wrapper.vm.table.records = []
    wrapper.vm.offset = offset
    wrapper.vm.limit = limit
    wrapper.vm.assignRecords(mock)
    expect(1).toEqual(wrapper.vm.table.records[0].id)
  })

  /**
   * @test
   */
  it('method assignRecordsFinish', () => {
    wrapper.vm.offset = 1
    wrapper.vm.assignRecordsFinish()
    expect(false).toEqual(wrapper.vm.table.loading)
  })

  /**
   * @test
   */
  it('method fetchRecords', async () => {
    wrapper.vm.table.records = []
    wrapper.vm.offset = offset
    wrapper.vm.limit = limit
    await wrapper.vm.fetchRecords()
    expect(mock.orders).toEqual(wrapper.vm.table.records)
  })

  /**
   * @test
   */
  it('method paginationUpdateOffset', async () => {
    wrapper.vm.table.records = []
    wrapper.vm.offset = offset
    wrapper.vm.limit = limit
    await wrapper.vm.paginationUpdateOffset({ offset })
    expect(mock.orders).toEqual(wrapper.vm.table.records)
  })

  /**
   * @test
   */
  it('method paginationToTop', () => {
    wrapper.vm.paginationToTop()
    expect(1).toEqual(1)
  })

  /**
   * @test
   */
  it('method loadingShow', () => {
    wrapper.vm.loadingShow()
    expect(true).toEqual(wrapper.vm.table.loading)
  })

  /**
   * @test
   */
  it('method loadingHide', () => {
    wrapper.vm.loadingHide()
    expect(false).toEqual(wrapper.vm.table.loading)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
