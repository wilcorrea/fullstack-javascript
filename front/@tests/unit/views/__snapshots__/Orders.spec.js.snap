// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders should render content correctly 1`] = `
<div class="app-layout__page Orders">
  <h1>Pedidos</h1>
  <app-table-stub columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" records="" loader="<div class=&quot;ld ld-ring ld-vortex-in&quot;></div>" presshold="true"></app-table-stub>
  <app-pagination-stub offset="0" limit="5"></app-pagination-stub>
</div>
`;
